File: codewars/4kyu/twice-linear/dduck-twice-linear.py
twice-linear# twice-linear
# https://www.codewars.com/kata/5672682212c8ecf83e000050

from unittest import TestCase


def dbl_linear(n):
    u = [1]
    pt2 = 0
    pt3 = 0
    i = 1

    while i <= n:
        u.append(min(2 * u[pt2] + 1, 3 * u[pt3] + 1))
        if u[i] == 2 * u[pt2] + 1:
            pt2 += 1
        if u[i] == 3 * u[pt3] + 1:
            pt3 += 1
        i += 1

    return u[n]


# it's imitation of JavaScript best practice


TestCase().assertEqual(dbl_linear(10), 22)
TestCase().assertEqual(dbl_linear(20), 57)
TestCase().assertEqual(dbl_linear(30), 91)
TestCase().assertEqual(dbl_linear(50), 175)

File: codewars/4kyu/twice-linear/dduck-twice-linear.js
// twice-linear
// https://www.codewars.com/kata/5672682212c8ecf83e000050

let sequence = [1];

for (let i = 0; i < 300000; i += 1) {
  sequence.push((sequence[i] * 2) + 1);
  sequence.push((sequence[i] * 3) + 1);
}
sequence = [...new Set(sequence)].sort((a, b) => a - b);

const dblLinear = (n) => sequence[n];

console.log(dblLinear(10), 22);
console.log(dblLinear(20), 57);
console.log(dblLinear(30), 91);
console.log(dblLinear(50), 175);
console.log(dblLinear(100), 447);
console.log(dblLinear(100000), 447);

// -------------------------
// best practice in CodeWars
// -------------------------
//
// function formatDuration (seconds) {
//   var time = { year: 31536000, day: 86400, hour: 3600, minute: 60, second: 1 },
//     res = [];
//
//   if (seconds === 0) return 'now';
//
//   for (var key in time) {
//     if (seconds >= time[key]) {
//       var val = Math.floor(seconds/time[key]);
//       res.push(val += val > 1 ? ' ' + key + 's' : ' ' + key);
//       seconds = seconds % time[key];
//     }
//   }
//
//   return res.length > 1 ? res.join(', ').replace(/,([^,]*)$/,' and'+'$1') : res[0]
// }
